spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -350,94 +442,57 @@ Access control on counterparty should inform choice of timeout values, i.e. time
-
-    // both channel ends must be mutually compatible.
-    // this means that the ordering must be the same and 
-    // any future introduced fields that must be compatible
-    // should also be checked
-    abortTransactionUnless(counterpartyChannel.ordering != proposedUpgradeChannel.ordering)
-    
-    // construct upgradeTimeout so it can be verified against counterparty state
-    upgradeTimeout = UpgradeTimeout{
-        timeoutHeight: timeoutHeight,
-        timeoutTimestamp: timeoutTimestamp,
-    }
-
-    // get underlying connection for proof verification
-    connection = getConnection(currentChannel.connectionIdentifier)
-
-    // verify proofs of counterparty state
-    abortTransactionUnless(verifyChannelState(connection, proofHeight, proofChannel, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
-    abortTransactionUnless(verifyChannelUpgradeTimeout(connection, proofHeight, proofUpgradeTimeout, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, upgradeTimeout))
-    abortTransactionUnless(verifyUpgradeSequence(connection, proofHeight, proofUpgradeSequence, currentChannel.counterpartyPortIdentifier,
-    currentChannel.counterpartyChannelIdentifier, counterpartySequence))
-
-    // get current sequence on this channel
-    currentSequence = provableStore.get(channelUpgradeSequencePath(portIdentifier, channelIdentifier))
-
-    // move upgrade sequence to fresh sequence for new upgrade attempt
-    if channel.state === types.OPEN {
-        channel.upgradeSequence = channel.upgradeSequence + 1
-
+    // if OPEN, then initialize handshake with upgradeFields
+    // otherwise, assert that the upgrade fields are the same for crossing-hellos case
+    if currentChannel.state == OPEN {
         // if the counterparty sequence is greater than the current sequence, we fast forward to the counterparty sequence
         // so that both channel ends are using the same sequence for the current upgrade
+        // initUpgradeChannelHandshake will increment the sequence so after that call
+        // both sides will have the same upgradeSequence
         if counterpartyUpgradeSequence > channel.upgradeSequence {
-            channel.upgradeSequence = counterpartyUpgradeSequence
+            channel.upgradeSequence = counterpartyUpgradeSequence - 1
         }
 
-        // this is first message in upgrade handshake on this chain so we must store original channel in restore channel path
-        // in case we need to restore channel later.
-        privateStore.set(channelRestorePath(portIdentifier, channelIdentifier), currentChannel)
-        provableStore.set(channelPath(portIdentifier, channelIdentifier), channel)
+        initUpgradeChannelHandshake(portIdentifier, channelIdentifier, upgradeFields, counterpartyUpgrade.timeout)
+    } else if currentChannel.state == INITUPGRADE {
+        existingUpgrade = publicStore.get(channelUpgradePath)
```

Comment:
do we expect the upgrade sequence to be fastforwarded in this case as well if the counterparty upgrade sequence is higher in crossing hellos?